docs/reference/scripting/using.asciidoc

```diff
@@ -1,425 +1,471 @@
+value is evaluated during an operation. Each type belongs to one of the
+following categories: primitive, reference, or dynamic.
+
+[[script-type-primitive]]
+==== Primitive types
+A primitive type holds a singular piece of data. Examples of
+{painless}/painless-types.html#primitive-types[available primitive types] are
+`byte`, `int`, and `long`. Primitive types can also specify the size and type
+of a variable value. For example, an `int` is 32-bit and stores whole numbers,
+whereas a `double` is 64-bit and stores fractional numbers.
+
+The following example declares three primitive types:
 
 [source,console]
---------------------------------------------------
-GET my-index-000001/_search
+----
+POST /_scripts/painless/_execute
 {
-  "query": {
-    "script_score": {
-      "query": {
-        "match": {
-            "message": "some message"
-        }
-      },
-      "script": {
-        "id": "calculate-score",
-        "params": {
-          "my_modifier": 2
-        }
-      }
-    }
+  "script": {
+    "source": """
+        int my_int; <1>
+        my_int = 1; <2>
+        double my_double = 2.0; <3>
+    """
   }
 }
---------------------------------------------------
-// TEST[continued]
-
-And deleted with:
+----
+<1> Primitive `int` type declaration
+<2> Primitive `int` type assignment with an `int` constant
+<3> Primitive `double` type declaration with assignment of a `double` constant
+
+You can declare a primitive type <<script-spec-variables,variable>> or access a
+primitive type member field from a reference type instance, and assign it a
+primitive type value for evaluation during later operations.
+
+Use the field access operator or method call operator on a primitive type value
+to force evaluation as its corresponding reference type value.
+
+[[script-type-reference]]
+==== Reference types
+A reference type is a complex structure that can represent multiple pieces of
+data and logic to manipulate that data. Reference types hold references to
+objects and provide a way to access them. You define reference types as part of
+the API for scripts.
+
+Reference types have methods and fields available to modify internal state. You
+cannot define new types, but can call any of the available types in the API's
+allowlist. For example, let's assume that the allowlist contains an `Example`
+reference type with the following definition:
+
+[source,java]
+----
+class Example {
+   int my_int;
+   double my_double;
+   def add(); <1>
+}
+----
+<1> This operation adds together the `my_int` and `my_double` variables.
 
-[source,console]
------------------------------------
-DELETE _scripts/calculate-score
------------------------------------
-// TEST[continued]
+To use the `Example` reference type, you create a new instance of `Example`
+that we'll name `my_example`. To call the `Example` method, include the
+{painless}/painless-operators-reference.html[method call operator] `()`, which
```

Comment:
Assuming `Example` had a constructor `Example(int my_int, double my_double)` this would be `new Example(123, 4.5)`.  Omitting `new` is incorrect and referring to calling the constructor as a using the `method call operator` is confusing.

Reply:
Thank you for clarifying and pointing out my errors. Is this statement more clear and accurate?

> To use the `Example` reference type, you create a new instance of `Example`
> that we'll name `my_example`. Assuming that `Example` has a constructor of
> `Example(int my_int, double my_double)`, you call `new Example()` as shown
> in the following example.
> 
> You can then define this new instance. The following example declares two
> primitive types named `my_int` and `my_double`, assigns them each a value, and
> includes them in the `my_example` reference type.
> 
> ```
> Example my_example = new Example()
> my_example.my_int = 1;
> my_example.my_double = 2;
> double my_add = my_example.add();
> ```

Reply:
Why not change `Example` to `HashMap`?  Its a common data structure and leads to more useful examples.

Backing up, who is the target audience for this section?  People with zero programming experience?  People with "a little" programming experience?

Reply:
Realistically, people with a little programming experience.

Reply:
A discussion of reference types isn't relevant for that audience.

The relevant information is that *mutable Objects* accessed through different variables change the underlying object.

How about an example with one of the common container types, `ArrayList` or `HashMap`? Create one instance, assign the reference to a difference variable, change it in via variable, observe the change in another`

```
Map mapA = new HashMap(); // mutable reference type
mapA.put("foo", 123);
Map mapB = new HashMap();
mapB.put("bar", 456);

Map mapRef = mapA;
mapRef.get("foo"); // Returns 123
mapRef.get("bar"); // Returns null, bar does not exist in this map

mapRef = mapB;
mapRef.get("foo"); // Returns null, foo does not exist in this map
mapRef.get("bar"); // Returns 456
```